src/common/com/intellij/plugins/haxe/util/HaxeImportUtil.java

```diff
@@ -17,311 +18,126 @@
  */
 package com.intellij.plugins.haxe.util;
 
-import com.intellij.openapi.util.Condition;
 import com.intellij.plugins.haxe.lang.psi.*;
+import com.intellij.plugins.haxe.model.HaxeImportModel;
 import com.intellij.psi.PsiElement;
 import com.intellij.psi.PsiFile;
-import com.intellij.util.containers.ContainerUtil;
+import com.intellij.psi.PsiPackage;
+import com.intellij.util.SmartList;
+import com.intellij.util.containers.MultiMap;
 import org.jetbrains.annotations.NotNull;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
+import java.util.stream.Collectors;
 
-/**
- * Created by fedorkorotkov.
- */
 public class HaxeImportUtil {
-  public static List<HaxeImportStatementRegular> findUnusedImports(PsiFile file) {
-    final List<HaxeClass> classesInFile = getReferencedClasses(file);
+  public static List<HaxeImportStatement> findUnusedImports(PsiFile file) {
+    final Collection<PsiElement> externalReferences = getExternalReferences(file);
 
-    List<HaxeImportStatementRegular> filteredUsefulImports = new ArrayList<HaxeImportStatementRegular>();
+    List<HaxeImportStatement> allImportStatements = ((HaxeFile)file).getImportStatements();
 
-    List<HaxeImportStatementRegular> allImportStatements = UsefulPsiTreeUtil.getAllImportStatements(file);
-    final List<HaxeImportStatementRegular> usefulImportStatements = ContainerUtil.findAll(allImportStatements, new Condition<HaxeImportStatementRegular>() {
-      @Override
-      public boolean value(HaxeImportStatementRegular statement) {
-        final HaxeImportStatementRegular regularImport = statement;
-        if(regularImport != null) {
-          final HaxeReferenceExpression referenceExpression = regularImport.getReferenceExpression();
+    final boolean hasWildcards = allImportStatements.stream().anyMatch(statement -> statement.getModel().hasWildcard());
 
-          if (null == referenceExpression) {
-            // It's an incomplete statement.  It can't be useful.
-            return false;
-          }
+    List<HaxeImportStatement> usefulStatements = allImportStatements
+      .stream()
+      .filter(statement -> externalReferences
+        .stream()
+        .anyMatch(referencedElement -> isStatementExposesReference(statement, referencedElement)))
+      .distinct()
+      .collect(Collectors.toList());
 
-          if (referenceExpression.resolve() == null) {
-            // don't know for sure
-            return true;
-          }
-        }
-        for (HaxeClass haxeClass : classesInFile) {
-          if (UsefulPsiTreeUtil.importStatementForClass(statement, haxeClass)) {
-            return true;
-          }
-        }
-        return false;
-      }
-    });
+    if (hasWildcards) {
+      removeUnusedWildcards(externalReferences, usefulStatements);
+    }
 
+    usefulStatements.forEach(allImportStatements::remove);
```

Comment:
Well this method was used deliberately, all because of equality check of removeAll. It check's not by reference equality but through `equals` + `hashCode` that isn't proper solution. We must to remove imports only from the list, using of removeAll will remove all duplicated import statements including that should remain.

Reply:
If that's the case, then add a comment, and then use an iterator explicitly, calling iterator.remove() to remove the element.  If you don't, then there will be subtle bugs because forEach() _is_ using an iterator and remove() makes it unstable (meaning the iterator doesn't advance properly if it doesn't know about the remove()).

Reply:
In essence, you're depending upon the underlying implementation of the iterator without it being specified in the contract.

Reply:
I think you are mistaken, though.  Both ArrayList.remove() and ArrayList.removeAll() use the same equality check (element.equals()) to determine if the element should be removed.  (And Collectors.toList() allocates an ArrayList.)

Reply:
Well, you can just check it, i've used this method only after using of removeAll, because of reasons described previously.

Reply:
I did go read the code for ArrayList to verify that they both use `equals()`.  I must have missed something then, if you tested it and found that this was the only workable solution.

Reply:
I figured this out:  Remove only removes the first object that matches via `equals()`, while removeAll removes everything that matches.  They both use the same algorithm to test for equality.  With the current code, the element removed will always be the first natural match (from the beginning of the array), not necessarily the actual object.  For the purposes of de-duping the imports, it works fine.  As long as the results are sorted, you'll likely never notice the difference.

Also, I realized (and I'm not sure why I didn't previously) that the iterated list and the affected list are different, so there is no issue with stability.